### ISSUE TITLE
No more error output but raise of error when loading input objects that have arguments with loads:

### ISSUE THREAD

**Issue by butsjoh** [rmosolgo/graphql-ruby#3746]
**Describe the bug**

No more error output but raise of error when loading input objects of different types. We think this is related to https://github.com/rmosolgo/graphql-ruby/pull/3682

Somehow the code is now raising "GraphQL::LoadApplicationObjectFailedError:" instead of nicely adding an error to the grapqhl response. This change since we upgraded to 1.13.0. I did not find anything in the documentation hinting at what the actual change is related to the merged PR. We are using the relay node resolving to let the objects load.  

**Versions**

`graphql` version: 1.13.0
`graphql` version before upgrade: 1.12.16

**GraphQL schema**

```ruby
module PublicGraph
  module Types
    class SomeInputType < GraphQL::Schema::InputObject
      graphql_name 'SomeInput'

      description "Input values to identify a member as ID or Email string"

      argument :member_id, ID, loads: Types::MemberType, required: false, as: :member
      argument :email, Types::EmailType, required: false

      validates required: { one_of: [:member, :email] }
    end
  end
end
```

```ruby

module PublicGraph
  module Resolvers
    class CreateMembershipsMutation < PublicGraph::BaseMutation

      description 'Creates a membership'

      argument :new_members, [Types:: SomeInputType], required: true, default_value: []

     ...
```

```ruby
module PublicGraph
  module Types
    class MemberType < GraphQL::Schema::Object
      implements GraphQL::Types::Relay::Node

      graphql_name 'Member'
      global_id_field :id

...
```

**GraphQL query**

Example GraphQL query and response (if query execution is involved)

```graphql

<<~GRAPHQL
    mutation createMembershipsMutation($newMembers: [SomeInput!]!) {
      createMemberships(
        input: {
          newMembers: $newMembers,
        }
      ) {
       ...
      }
    }
```

**Expected behavior**

Before it would not raise hard but just give a clean result which contained the "No object found for" in the errors output but now it is just raising hard.

```ruby
      result = PublicGraph::Schema.execute(
        "
     mutation createMembershipsMutation($newMembers: [SomeInput!]!) {
        createMemberships(
          input: {
            newMembers: $newMembers,
          }
        ) {
         ...
        }
      }
        ",
        variables: {
          "newMembers": [
            {memberId: member2.graphql_id},
            {memberId: member3.graphql_id},
            {memberId: member4.graphql_id},
            {email: Faker::Internet.unique.email},
          ]
        },
        context: {
          ...
        }
      )

      expect(result.dig('data', 'createMemberships')).to be_nil
      expect(result.dig('errors', 0, 'message')).to include('No object found for `memberId:')
```

**Actual behavior**

<img width="1205" alt="Screenshot 2021-12-01 at 10 31 13" src="https://user-images.githubusercontent.com/96848/144208750-6b8f598e-a7f0-4199-9586-d8e0db47913c.png">

**Comment by butsjoh** [comment]
So i guess it is related to this change: https://github.com/rmosolgo/graphql-ruby/pull/3689

I still find it weird that the behaviour all of a sudden changed ...

**Comment by butsjoh** [comment]
Tried adding this to my mutation and the input object but still raising an error

```
      def load_application_object_failed(_)
        nil
      end
```

**Comment by butsjoh** [comment]
This seems to be really specific to InputObjects cause if a same loads happens in a normal mutation it does get to the load_application_object_failed method  but in case of the input object it does not.

**Comment by rmosolgo** [comment]
Thanks for reporting this issue with so much detail. It's likely a bug that I introduced when "improving" the `loads:` behavior for input objects 🤦 . I'll take a look and follow up here soon!

**Comment by rmosolgo** [comment]
I tried to replicate this bug in a test in https://github.com/rmosolgo/graphql-ruby/commit/c4c587741b9b09efea09663b102e955a1a231f4b, but the test passed, so I must be missing something. 

One interesting thing about the error is that the raised error is coming from error handling code, trying to get `.keyword_arguments` from a `GraphQL::Execution::Lazy` instance. That means that the object was loaded from ID by batch loading, I'd guess. 

A couple questions: 

- Can you make your test `rescue StandardError => err` then `puts err.cause.backtrace`? That would give more information about where the `LoadApplicationObjectFailedError` came from 
- Do you have any `rescue_from` handlers in your schema? Would you mind sharing `rescue_from(...)`, without the handler block? It seems like GraphQL-Ruby has matched `LoadApplicationObjectFailedError` to one of those handlers. (This might be the _right_ behavior -- but maybe GraphQL-Ruby used to rescue those errors internally better.) 
- When it's loading this object by ID, does it find something in the database at all, or does `Schema.object_from_id` return `nil`? 

🤔  I'd like to figure out where this is getting raised _from_ -- let me know what you find!

**Comment by butsjoh** [comment]
I will get you the information tomorrow cause i already winded down my dev env :)

What i can already confirm is that indeed that id does not return anything from the db and that is what the test is actually testing :) We do use graphql-batch so on node resolving we do call the load(id) method on a GraphQL::Batch::Loader instance. Maybe that cause the result to still be lazy and maybe that is not handled correctly by input object. We don't have this problem if we use loads: for arguments on mutations.

**Comment by butsjoh** [comment]
Regarding rescue_from we have the following:

```
    rescue_from(StandardError) do |err, _obj, _args, _ctx, _field|
      case err
      when ::ActionPolicy::Unauthorized
        raise PublicGraph::NotAuthorized
      else
        raise err
      end
```

**Comment by butsjoh** [comment]
Could not wait and still got the error result:

```ruby
 ["/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema/member/has_arguments.rb:321:in `load_application_object_failed'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema/member/has_arguments.rb:280:in `block in authorize_application_object'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:118:in `after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:115:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:43:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `public_send'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `sync_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:113:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block (2 levels) in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `map'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:65:in `block in then'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `public_send'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `sync_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:113:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block (2 levels) in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `map'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:65:in `block in then'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `public_send'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `sync_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:113:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block (2 levels) in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `map'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:65:in `block in then'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `public_send'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `sync_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:113:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `public_send'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `sync_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:113:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block (2 levels) in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `map'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:73:in `block in all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:65:in `block in then'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `public_send'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:129:in `sync_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter/runtime.rb:875:in `block (3 levels) in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/tracing.rb:66:in `trace'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter/runtime.rb:874:in `block (2 levels) in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/query.rb:366:in `block in with_error_handling'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/errors.rb:107:in `with_error_handling'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/query.rb:365:in `with_error_handling'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter/runtime.rb:872:in `block in after_lazy'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/lazy.rb:41:in `value'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter/resolve.rb:47:in `resolve'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter/resolve.rb:10:in `block in resolve_all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/dataloader/null_dataloader.rb:17:in `append_job'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter/resolve.rb:10:in `resolve_all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter.rb:94:in `block in sync_lazies'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/tracing.rb:66:in `trace'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter.rb:93:in `sync_lazies'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/interpreter.rb:54:in `finish_multiplex'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:111:in `block in run_as_multiplex'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/dataloader/null_dataloader.rb:17:in `append_job'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:110:in `run_as_multiplex'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:63:in `block (2 levels) in run_queries'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:208:in `block in instrument_and_analyze'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:29:in `block (2 levels) in apply_instrumenters'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:46:in `block (2 levels) in each_query_call_hooks'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:41:in `each_query_call_hooks'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:45:in `block in each_query_call_hooks'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:44:in `each_query_call_hooks'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:27:in `block in apply_instrumenters'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/instrumentation.rb:26:in `apply_instrumenters'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:187:in `instrument_and_analyze'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:62:in `block in run_queries'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/tracing.rb:66:in `trace'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:60:in `run_queries'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/execution/multiplex.rb:50:in `run_all'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:1744:in `multiplex'",
 "/bundle/ruby/3.0.0/gems/graphql-1.13.0/lib/graphql/schema.rb:1715:in `execute'",
 "/graph/spec/graph/spaces/create_memberships_mutation_spec.rb:270:in `block (3 levels) in <top (required)>'"
...
]
```

**Comment by rmosolgo** [comment]
thanks again for those details. I'm guessing that previously, this error was passed through your handler to the `raise err` in it. But I added some some stuff to GraphQL-Ruby's error handler which was breaking _before_ reaching your handler there. The solution is to rescue GraphQL-Ruby's own errors _before_ engaging the error handler code, avoiding the error you posted above. Fix in the works at #3749

**Comment by butsjoh** [comment]
I just confirmed that indeed in 1.12.16 i see the GraphQL::LoadApplicationObjectFailedError being reraised in our rescue_from block which no longer happens in 1.13.0.
